### DISCUSSION TITLE
How does one troubleshoot JSON errors?

### DISCUSSION THREAD

**Discussion by chewxy** [joaotavora/eglot#1358] [Q&A]
Hi, I'm having some trouble with `gopls` which is the language server for Go. The error/warning I keep getting is 

```
Warning (eglot): could not parse JSON stream: "invalid token near 'go'", "<callback>", 1, 2, 2
```

I had turned on tracing and debugging  on `gopls` and so I am able to see every JSON message received. However, I note from eglot's point of view, eglot simply sends the handshake message and doesn't do anything with it:

```
jsonrpc] e[06:32:49.687] <-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2024/01/29 06:32:49 Handshake session update\n\tupdate_session=4\n\tdebug_address=\"\"\n\tlogfile=\"/var/folders/4v/79zdj6w13cj8d8_xy7zxpbq40000gq/T/gopls-17337.log\"\n\tserver=\"1\"\n\tgopls_path=\"/Users/chewxy/go/bin/gopls\"\n"}}
```

And then sends the warning. Eglot doesn't seem to have sent `initialize - (1)` to `gopls`. I don't see the typical `initialize (1)` message on gopls' log. The entirety of the log file above is as follows: 

```
[Trace - 06:32:49.686 AM] Received notification 'window/logMessage'.
Params: {"type":3,"message":"2024/01/29 06:32:49 Handshake session update\n\tupdate_session=4\n\tdebug_address=\"\"\n\tlogfile=\"/var/folders/4v/79zdj6w13cj8d8_xy7zxpbq40000gq/T/gopls-17337.log\"\n\tserver=\"1\"\n\tgopls_path=\"/Users/xuanyi.chew/go/bin/gopls\"\n"}


```

Any help? It seems to me gopls may be sending something (but in that case the sending should show up on gopls' log)? or Eglot may be trying to send an `initialize (1)` to gopls but somehow has to parse a json somewhere.

I'm on 

```
GNU Emacs 30.0.50 (build 2, aarch64-apple-darwin23.2.0, NS appkit-2487.30 Version 14.2 (Build 23C64)) of 2024-01-25
```

Eglot's on version 1.16 (per pkg-info-version-info)

JSONRPC is on version 1.0.24 (per pkg-info-version-info)

**Comment by joaotavora** [comment]
>  However, I note from eglot's point of view, eglot simply sends the handshake message and doesn't do anything with it:

How are you accessing this "Eglot point of view".  Is it via `M-x eglot-events-buffer`?  If so, are you sure you're viewing the full buffer?

For reference, here's the full transcript of a the start of `gopls` session on a simple "hello world" project, grabbed from `M-x eglot-events-buffer`.
<details><summary>transcript</summary>
<pre>
[jsonrpc] D[18:00:42.904] Running language server: gopls
[jsonrpc] e[18:00:42.905] --> initialize[1] {"jsonrpc":"2.0","id":1,"method":"initialize","params":{"processId":4673,"clientInfo":{"name":"Eglot","version":"1.17"},"rootPath":"/home/capitaomorte/Source/Go/hello/","rootUri":"file:///home/capitaomorte/Source/Go/hello","initializationOptions":{},"capabilities":{"workspace":{"applyEdit":true,"executeCommand":{"dynamicRegistration":false},"workspaceEdit":{"documentChanges":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":false},"configuration":true,"workspaceFolders":true},"textDocument":{"synchronization":{"dynamicRegistration":false,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":false,"completionItem":{"snippetSupport":true,"deprecatedSupport":true,"resolveSupport":{"properties":["documentation","details","additionalTextEdits"]},"tagSupport":{"valueSet":[1]}},"contextSupport":true},"hover":{"dynamicRegistration":false,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":false,"signatureInformation":{"parameterInformation":{"labelOffsetSupport":true},"documentationFormat":["markdown","plaintext"],"activeParameterSupport":true}},"references":{"dynamicRegistration":false},"definition":{"dynamicRegistration":false,"linkSupport":true},"declaration":{"dynamicRegistration":false,"linkSupport":true},"implementation":{"dynamicRegistration":false,"linkSupport":true},"typeDefinition":{"dynamicRegistration":false,"linkSupport":true},"documentSymbol":{"dynamicRegistration":false,"hierarchicalDocumentSymbolSupport":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"documentHighlight":{"dynamicRegistration":false},"codeAction":{"dynamicRegistration":false,"resolveSupport":{"properties":["edit","command"]},"dataSupport":true,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}},"isPreferredSupport":true},"formatting":{"dynamicRegistration":false},"rangeFormatting":{"dynamicRegistration":false},"rename":{"dynamicRegistration":false},"inlayHint":{"dynamicRegistration":false},"publishDiagnostics":{"relatedInformation":false,"codeDescriptionSupport":false,"tagSupport":{"valueSet":[1,2]}}},"window":{"showDocument":{"support":true},"workDoneProgress":true},"general":{"positionEncodings":["utf-32","utf-8","utf-16"]},"experimental":{}},"workspaceFolders":[{"uri":"file:///home/capitaomorte/Source/Go/hello","name":"~/Source/Go/hello/"}]}}
[jsonrpc] e[18:00:43.471] <-- initialize[1] {"jsonrpc":"2.0","result":{"capabilities":{"textDocumentSync":{"openClose":true,"change":2,"save":{}},"completionProvider":{"triggerCharacters":["."]},"hoverProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",","]},"definitionProvider":true,"typeDefinitionProvider":true,"implementationProvider":true,"referencesProvider":true,"documentHighlightProvider":true,"documentSymbolProvider":true,"codeActionProvider":{"codeActionKinds":["quickfix","refactor.extract","refactor.inline","refactor.rewrite","source.fixAll","source.organizeImports"]},"codeLensProvider":{},"documentLinkProvider":{},"workspaceSymbolProvider":true,"documentFormattingProvider":true,"renameProvider":true,"foldingRangeProvider":true,"selectionRangeProvider":true,"executeCommandProvider":{"commands":["gopls.add_dependency","gopls.add_import","gopls.add_telemetry_counters","gopls.apply_fix","gopls.change_signature","gopls.check_upgrades","gopls.edit_go_directive","gopls.fetch_vulncheck_result","gopls.gc_details","gopls.generate","gopls.go_get_package","gopls.list_imports","gopls.list_known_packages","gopls.maybe_prompt_for_telemetry","gopls.mem_stats","gopls.regenerate_cgo","gopls.remove_dependency","gopls.reset_go_mod_diagnostics","gopls.run_go_work_command","gopls.run_govulncheck","gopls.run_tests","gopls.start_debugging","gopls.start_profile","gopls.stop_profile","gopls.test","gopls.tidy","gopls.toggle_gc_details","gopls.update_go_sum","gopls.upgrade_dependency","gopls.vendor","gopls.workspace_stats"]},"callHierarchyProvider":true,"semanticTokensProvider":{"legend":{"tokenTypes":[],"tokenModifiers":[]},"range":true,"full":true},"inlayHintProvider":{},"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":"workspace/didChangeWorkspaceFolders"}}},"serverInfo":{"name":"gopls","version":"{\"GoVersion\":\"go1.21.5\",\"Path\":\"golang.org/x/tools/gopls\",\"Main\":{\"Path\":\"golang.org/x/tools/gopls\",\"Version\":\"v0.14.2\",\"Sum\":\"h1:sIw6vjZiuQ9S7s0auUUkHlWgsCkKZFWDHmrge8LYsnc=\",\"Replace\":null},\"Deps\":[{\"Path\":\"github.com/BurntSushi/toml\",\"Version\":\"v1.2.1\",\"Sum\":\"h1:9F2/+DoOYIOksmaJFPw1tGFy1eDnIJXg+UHjuD8lTak=\",\"Replace\":null},{\"Path\":\"github.com/google/go-cmp\",\"Version\":\"v0.5.9\",\"Sum\":\"h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=\",\"Replace\":null},{\"Path\":\"github.com/sergi/go-diff\",\"Version\":\"v1.1.0\",\"Sum\":\"h1:we8PVUC3FE2uYfodKH/nBHMSetSfHDR6scGdBi+erh0=\",\"Replace\":null},{\"Path\":\"golang.org/x/exp/typeparams\",\"Version\":\"v0.0.0-20221212164502-fae10dda9338\",\"Sum\":\"h1:2O2DON6y3XMJiQRAS1UWU+54aec2uopH3x7MAiqGW6Y=\",\"Replace\":null},{\"Path\":\"golang.org/x/mod\",\"Version\":\"v0.14.0\",\"Sum\":\"h1:dGoOF9QVLYng8IHTm7BAyWqCqSheQ5pYWGhzW00YJr0=\",\"Replace\":null},{\"Path\":\"golang.org/x/sync\",\"Version\":\"v0.4.0\",\"Sum\":\"h1:zxkM55ReGkDlKSM+Fu41A+zmbZuaPVbGMzvvdUPznYQ=\",\"Replace\":null},{\"Path\":\"golang.org/x/sys\",\"Version\":\"v0.14.0\",\"Sum\":\"h1:Vz7Qs629MkJkGyHxUlRHizWJRG2j8fbQKjELVSNhy7Q=\",\"Replace\":null},{\"Path\":\"golang.org/x/telemetry\",\"Version\":\"v0.0.0-20231114163143-69313e640400\",\"Sum\":\"h1:brbkEFfGwNGAEkykUOcryE/JiHUMMJouzE0fWWmz/QU=\",\"Replace\":null},{\"Path\":\"golang.org/x/text\",\"Version\":\"v0.13.0\",\"Sum\":\"h1:ablQoSUd0tRdKxZewP80B+BaqeKJuVhuRxj/dkrun3k=\",\"Replace\":null},{\"Path\":\"golang.org/x/tools\",\"Version\":\"v0.14.1-0.20231114185516-c9d3e7de13fd\",\"Sum\":\"h1:Oku7E+OCrXHyst1dG1z10etCTxewCHXNFLRlyMPbh3w=\",\"Replace\":null},{\"Path\":\"golang.org/x/vuln\",\"Version\":\"v1.0.1\",\"Sum\":\"h1:KUas02EjQK5LTuIx1OylBQdKKZ9jeugs+HiqO5HormU=\",\"Replace\":null},{\"Path\":\"honnef.co/go/tools\",\"Version\":\"v0.4.5\",\"Sum\":\"h1:YGD4H+SuIOOqsyoLOpZDWcieM28W47/zRO7f+9V3nvo=\",\"Replace\":null},{\"Path\":\"mvdan.cc/gofumpt\",\"Version\":\"v0.4.0\",\"Sum\":\"h1:JVf4NN1mIpHogBj7ABpgOyZc65/UUOkKQFkoURsz4MM=\",\"Replace\":null},{\"Path\":\"mvdan.cc/xurls/v2\",\"Version\":\"v2.4.0\",\"Sum\":\"h1:tzxjVAj+wSBmDcF6zBB7/myTy3gX9xvi8Tyr28AuQgc=\",\"Replace\":null}],\"Settings\":[{\"Key\":\"-buildmode\",\"Value\":\"exe\"},{\"Key\":\"-compiler\",\"Value\":\"gc\"},{\"Key\":\"DefaultGODEBUG\",\"Value\":\"panicnil=1\"},{\"Key\":\"CGO_ENABLED\",\"Value\":\"1\"},{\"Key\":\"CGO_CFLAGS\",\"Value\":\"\"},{\"Key\":\"CGO_CPPFLAGS\",\"Value\":\"\"},{\"Key\":\"CGO_CXXFLAGS\",\"Value\":\"\"},{\"Key\":\"CGO_LDFLAGS\",\"Value\":\"\"},{\"Key\":\"GOARCH\",\"Value\":\"amd64\"},{\"Key\":\"GOOS\",\"Value\":\"linux\"},{\"Key\":\"GOAMD64\",\"Value\":\"v1\"}],\"Version\":\"v0.14.2\"}"}},"id":1}
[jsonrpc] e[18:00:43.471] --> initialized {"jsonrpc":"2.0","method":"initialized","params":{}}
[jsonrpc] e[18:00:43.474] --> textDocument/didOpen {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"uri":"file:///home/capitaomorte/Source/Go/hello/hello.go","version":0,"languageId":"go","text":"package main\n\nimport \"fmt\"\n\nfunc main() {\n    \n}"}}}
[jsonrpc] e[18:00:43.477] --> workspace/didChangeConfiguration {"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}
[jsonrpc] e[18:00:43.481] --> textDocument/inlayHint[2] {"jsonrpc":"2.0","id":2,"method":"textDocument/inlayHint","params":{"textDocument":{"uri":"file:///home/capitaomorte/Source/Go/hello/hello.go"},"range":{"start":{"line":0,"character":0},"end":{"line":6,"character":1}}}}
[jsonrpc] e[18:00:43.482] <-- window/workDoneProgress/create[1] {"jsonrpc":"2.0","method":"window/workDoneProgress/create","params":{"token":"3721461203173796616"},"id":1}
[jsonrpc] e[18:00:43.482] --> window/workDoneProgress/create[1] {"jsonrpc":"2.0","id":1,"result":null}
[jsonrpc] e[18:00:43.482] <-- $/progress {"jsonrpc":"2.0","method":"$/progress","params":{"token":"3721461203173796616","value":{"kind":"begin","title":"Setting up workspace","message":"Loading packages..."}}}
[jsonrpc] e[18:00:43.504] <-- workspace/configuration[2] {"jsonrpc":"2.0","method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///home/capitaomorte/Source/Go/hello","section":"gopls"}]},"id":2}
[jsonrpc] e[18:00:43.505] --> workspace/configuration[2] {"jsonrpc":"2.0","id":2,"result":[null]}
[jsonrpc] e[18:00:43.541] <-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2024/01/28 18:00:43 go info for /home/capitaomorte/Source/Go/hello\n(go dir /home/capitaomorte/Source/Go/hello)\n(go version go version go1.21.6 linux/amd64)\n(valid build configuration = true)\n(build flags: [])\n(selected go env: [GO111MODULE=, GOCACHE=/home/capitaomorte/.cache/go-build, GOFLAGS=, GOMODCACHE=/home/capitaomorte/.golang/pkg/mod, GOPATH=/home/capitaomorte/.golang, GOPRIVATE=, GOROOT=/usr/lib/go, GOWORK=])\n\n"}}
[jsonrpc] e[18:00:43.705] <-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2024/01/28 18:00:43 go/packages.Load #1\n\tsnapshot=0\n\tdirectory=file:///home/capitaomorte/Source/Go/hello\n\tquery=[/home/capitaomorte/Source/Go/hello/... builtin]\n\tpackages=2\n"}}
[jsonrpc] e[18:00:43.705] <-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2024/01/28 18:00:43 go/packages.Load #1: updating metadata for 43 packages\n"}}
[jsonrpc] e[18:00:43.734] <-- $/progress {"jsonrpc":"2.0","method":"$/progress","params":{"token":"3721461203173796616","value":{"kind":"end","message":"Finished loading packages."}}}
[jsonrpc] e[18:00:43.734] <-- client/registerCapability[3] {"jsonrpc":"2.0","method":"client/registerCapability","params":{"registrations":[{"id":"workspace/didChangeWatchedFiles-0","method":"workspace/didChangeWatchedFiles","registerOptions":{"watchers":[{"globPattern":"**/*.{go,mod,sum,work}","kind":7}]}}]},"id":3}
[jsonrpc] e[18:00:43.745] --> client/registerCapability[3] {"jsonrpc":"2.0","id":3,"result":null}
[jsonrpc] e[18:00:43.749] <-- workspace/configuration[4] {"jsonrpc":"2.0","method":"workspace/configuration","params":{"items":[{"section":"gopls"}]},"id":4}
[jsonrpc] e[18:00:43.749] --> workspace/configuration[4] {"jsonrpc":"2.0","id":4,"result":[null]}
[jsonrpc] e[18:00:43.750] <-- workspace/configuration[5] {"jsonrpc":"2.0","method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///home/capitaomorte/Source/Go/hello","section":"gopls"}]},"id":5}
[jsonrpc] e[18:00:43.751] --> workspace/configuration[5] {"jsonrpc":"2.0","id":5,"result":[null]}
[jsonrpc] e[18:00:43.821] <-- textDocument/inlayHint[2] {"jsonrpc":"2.0","result":null,"id":2}
[jsonrpc] e[18:00:43.852] <-- textDocument/publishDiagnostics {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/capitaomorte/Source/Go/hello/hello.go","diagnostics":[{"range":{"start":{"line":2,"character":7},"end":{"line":2,"character":12}},"severity":1,"code":"UnusedImport","codeDescription":{"href":"https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#UnusedImport"},"source":"compiler","message":"\"fmt\" imported and not used","tags":[1]}]}}
[jsonrpc] e[18:00:43.983] --> textDocument/hover[3] {"jsonrpc":"2.0","id":3,"method":"textDocument/hover","params":{"textDocument":{"uri":"file:///home/capitaomorte/Source/Go/hello/hello.go"},"position":{"line":5,"character":4}}}
[jsonrpc] e[18:00:43.983] --> textDocument/documentHighlight[4] {"jsonrpc":"2.0","id":4,"method":"textDocument/documentHighlight","params":{"textDocument":{"uri":"file:///home/capitaomorte/Source/Go/hello/hello.go"},"position":{"line":5,"character":4}}}
[jsonrpc] e[18:00:43.984] --> textDocument/signatureHelp[5] {"jsonrpc":"2.0","id":5,"method":"textDocument/signatureHelp","params":{"textDocument":{"uri":"file:///home/capitaomorte/Source/Go/hello/hello.go"},"position":{"line":5,"character":4}}}
[jsonrpc] e[18:00:43.985] <-- textDocument/hover[3] {"jsonrpc":"2.0","result":null,"id":3}
[jsonrpc] e[18:00:43.985] <-- textDocument/documentHighlight[4] {"jsonrpc":"2.0","result":[],"id":4}
[jsonrpc] e[18:00:43.985] <-- textDocument/signatureHelp[5] {"jsonrpc":"2.0","result":null,"id":5}
[jsonrpc] e[18:00:43.985] <-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":1,"message":"2024/01/28 18:00:43 no signature help: cannot find an enclosing function\n\tposition=5:4\n"}}
</pre>
</details> 

I'm using a fairly recent build from Emacs master, which contains Eglot 1.17.  Gopls is  gopls version is
```
golang.org/x/tools/gopls v0.14.2
    golang.org/x/tools/gopls@v0.14.2 h1:sIw6vjZiuQ9S7s0auUUkHlWgsCkKZFWDHmrge8LYsnc=
```

**Comment by chewxy** [comment]
> Is it via `M-x eglot-events-buffer`?

Yep. It's via that. I'm not sure about the full buffer. What I'm doing is I have that buffer open in a window, and then I open a Go buffer in another window, then manually calling `M-x eglot`. It's a very short buffer, often looking something like this (this is with the debug mode turned off):

```
[jsonrpc] D[15:43:21.902] Running language server: gopls
[stderr]  
[stderr]  
[stderr]  nil
[stderr]  nil
[stderr]  Process EGLOT (gorgonia/(go-ts-mode go-mode go-dot-mod-mode go-dot-work-mode go-mod-ts-mode)) stderr finished
[jsonrpc] D[15:43:24.350] Running language server: gopls
[stderr]  
[stderr]  
[stderr]  nil
[stderr]  nil
[stderr]  Process EGLOT (gorgonia/(go-ts-mode go-mode go-dot-mod-mode go-dot-work-mode go-mod-ts-mode)) stderr<1> finished
[jsonrpc] D[15:45:04.563] Running language server: gopls
```

Attempting to get the stderr buffer with `M-x eglot-stderr-buffer` yields the following message: 

```
jsonrpc-error: jsonrpc-error: "No current JSON-RPC connection", (jsonrpc-error-code . -32603), (jsonrpc-error-message . "No current JSON-RPC connection")
```

**Comment by joaotavora** [comment]
Chewxy ***@***.***> writes:

>  Is it via M-x eglot-events-buffer?
>
> Yep. It's via that. I'm not sure about the full buffer. What I'm doing is I have that buffer open in a window, and then I open a Go buffer in another window,
> then manually calling M-x eglot. It's a very short buffer, often looking something like this (this is with the debug mode turned off):
>
> [jsonrpc] D[15:43:21.902] Running language server: gopls
> [stderr]  
> [stderr]  
> [stderr]  nil
> [stderr]  nil
> [stderr]  Process EGLOT (gorgonia/(go-ts-mode go-mode go-dot-mod-mode go-dot-work-mode go-mod-ts-mode)) stderr finished
> [jsonrpc] D[15:43:24.350] Running language server: gopls
> [stderr]  
> [stderr]  
> [stderr]  nil
> [stderr]  nil
> [stderr]  Process EGLOT (gorgonia/(go-ts-mode go-mode go-dot-mod-mode go-dot-work-mode go-mod-ts-mode)) stderr<1> finished
> [jsonrpc] D[15:45:04.563] Running language server: gopls

Maybe you ought to check your config, you might have a

  (setq eglot-events-buffer-size 0)

Somewhere in it.  So I'd try again with Emacs -Q and if it still fails
post more details.  On this side:

   emacs -Q hello/hello.go -f go-ts-mode -f eglot

works perfectly, both with Emacs 29 and Emacs 30.

   emacs -Q hello/hello.go -f package-initialize -f go-mode -f eglot

Also works perfectly.

**Comment by chewxy** [comment]
No joy unfortunately. On my machine somehow it's `eglot-events-buffer-config`, instead of `eglot-events-buffer-size`

```
eglot-events-buffer-config is a variable defined in ‘eglot.el’.

Its value is (:size 2000000 :format full)
```

**Comment by chewxy** [comment]
additional note: eglot seems to work fine with pylsp, and certain Go files (coincidentally in repositories with majority Python code).

My goal in asking the question of how to debug the json, is to see if particular settings would cause eglot to send/receive bad JSON. Unfortunately it's rather hard to do that when apparently no JSON is sent/received by either party.
